README.md
frontaccounting
===============

* This is an unofficial Git Repo for frontaccounting and it's extensions / utilities.
* Currently only the official v2.3.x branch is supported here.
* My mods and pending commits to the core are in the [FAMods](https://github.com/apmuthu/frontaccounting/blob/master/FAMods) folder and documented in the [FA 2.3 Support Wiki](https://github.com/apmuthu/frontaccounting/wiki).
* Column Widths for large transaction numbers in reports [discussed in the forum](http://frontaccounting.com/punbb/viewtopic.php?id=6456) can be [backported](https://github.com/apmuthu/frontac24/commit/08b81e2fe2536c7c9f7146184dfe4fc37b57c32d).
* DB Schema has been frozen for this version and any changes will break normal upgradeability.
* DB Schema can be optimised as in [FAMods/sql/alter_to_latest.sql](https://github.com/apmuthu/frontaccounting/blob/master/FAMods/sql/alter_to_latest2.3.sql)

Official Links:
===============
<ul>
<li>Home: http://www.frontaccounting.com</li>
<li>Releases: http://www.sf.net/projects/frontaccounting/files</li>
<li>Forums: http://frontaccounting.com/punbb</li>
<li>Wiki: http://frontaccounting.com/fawiki/</li>
<li>ERD in Wiki: http://frontaccounting.com/fawiki/index.php?n=Devel.ERDiagram23</li>
<li>Reports in Wiki: http://frontaccounting.com/fawiki/index.php?n=Help.ReportsAndAnalysis</li>
<li>Bugs: http://mantis.frontaccounting.com/</li>
<li>Extensions: http://anonymous:password@repo.frontaccounting.eu/2.3/</li> - updation stopped at the beginning of 2016 for this version.
</ul>

The official [FrontAccounting Source Tree](http://sourceforge.net/p/frontaccounting/git/ci/master/tree/) uses Git on SourceForge since 2015-03-29
and has been officially [mirrored on GitHub](https://github.com/FrontAccountingERP/FA) from 2015-03-29 onwards.

Caveats:
========
* This repo will be updated only when I have the time.
* It is meant to enable easy localised forking on GitHub from any release.
* It can be used to study differences for migration between versions where individual proficiency in using GIT is better than in using Mercurial.
* Like any development code repo, code here should be used with care.
* Do not be fooled that any snapshot is well tested stable release, while in many times it is not.
* Using full random snapshots of FA repo is officially unwelcome.
* FA Repo code is placed here for study, development and local deployment only especially where internet is not available.
* FA versions are not released too often in order to provide a minimal acceptable level of stability, which is important especially in accounting software.
* Any changes in the code that are not yet in the official repo is listed in [FAMods/CHANGELOG_apmuthu.txt](https://github.com/apmuthu/frontaccounting/blob/master/FAMods/CHANGELOG_apmuthu.txt).

Third Party Connects
====================
* <b>WordPress Connect</b>: [WP SVN](http://plugins.svn.wordpress.org/frontaccounting-connect/trunk/). The [original repo](https://github.com/wp-plugins/frontaccounting-connect) is empty now.
* Import active Frontaccounting items into wordpress as posts.


core/CHANGELOG.txt
2017-05-20 Joe Hunt           Sales type potential bug in tax_included (persistent on)
2017-05-20 Joe Hunt           Deleted doc/CHANGELOG.old.txt. Not needed any more.
2017-05-20 Joe Hunt           0003761: Reorder email is not sent for items used work order. Fixed
2017-05-12 Joe Hunt           Optimized Inventory Valuation Reports, 301 and 308.
2017-05-11 Joe Hunt           Customer/Supplier Balances Report | Allocated column not showing correctly. Fixed.
2017-03-19 Ap.Muthu           Entropy fix update. random_id() in main.inc wraps native uniqid()
2017-01-18 Ap.Muthu           Fixed an array_combine object issue.
2016-12-23 Ap.Muthu           Moved from FAMods: Fixed bug - submitting forms in FA 2.3.x when checkbox field not changed
2016-12-04 Joe Hunt           Rerun Item Sales Summary Report fix
2016-11-29 Joe Hunt           Fixed A Bug in Item Sales Summary Report
2016-11-29 Joe Hunt           0003667: supplier payment link after entering supplier invoice result. Fixed. 
2016-11-14 Janusz Dobrowolski Cleanup in sales_invoice_db.inc (obsolete parameters in update_debtor_trans_allocation calls)
2016-07-26 Joe Hunt           Dimension is missing in Supplier Invoice View, GL Rows.
2016-07-24 Ap.Muthu           Minor bugs in deleting fiscal years
2016-05-23 Joe Hunt           Tabbed dialogs allowed non access pages.
2016-05-23 Joe Hunt           Bloomberg Rate fix for change of wrapper from span to div.
2016-05-20 Janusz Dobrowolski Bank Accounts Edition fixes
2016-05-17 Janusz Dobrowolski de_DE and pt_PT installer translation typos fixed

2016-02-21 15:34:07 +0100 Joe Hunt 
Release 2.3.25.
M	version.php

2016-01-13 08:21:19 +0100 Joe Hunt 
Created new fiscal years in en_US-demo.sql
M	sql/en_US-demo.sql

2016-01-10 09:50:38 +0100 Janusz Dobrowolski 
Reverted invalid changes in manufacturing accounts back to cogs_account for BOM service items.
M	manufacturing/includes/db/work_order_issues_db.inc
M	manufacturing/includes/db/work_order_produce_items_db.inc
M	manufacturing/includes/db/work_orders_db.inc
M	manufacturing/includes/db/work_orders_quick_db.inc

2015-12-09 14:04:45 +0100 Joe Hunt 
Fixed a hardcoded amount_row with fixed 2 decimals. Removed and the default user_price_dec() is used.
M	purchasing/supplier_payment.php

2015-11-26 21:48:19 +0100 Janusz Dobrowolski 
Fixed es_MX installer translation file.
M	install/lang/es_MX/LC_MESSAGES/es_MX.mo
M	install/lang/es_MX/LC_MESSAGES/es_MX.po

2015-11-23 09:26:48 +0100 Joe Hunt 
No display of shipping if there are no shipping costs on sales documents or views.
M	reporting/rep107.php
M	reporting/rep109.php
M	reporting/rep110.php
M	reporting/rep111.php
M	reporting/rep113.php
M	sales/view/view_credit.php
M	sales/view/view_dispatch.php
M	sales/view/view_invoice.php
M	sales/view/view_sales_order.php

2015-11-14 22:26:25 +0100 Joe Hunt 
Rerun. Backport of fixing vertical alignment on date picker icon
M	reporting/includes/reports_classes.inc

2015-11-14 09:34:54 +0100 Joe Hunt 
Accounts entry. We only accept 0-9 and a dot if $accounts_alpha = 0 (default)
M	gl/manage/gl_accounts.php

2015-11-13 11:54:10 +0100 Joe Hunt 
Backport of fixing vertical alignment on date picker icon
M	includes/ui/ui_input.inc
M	reporting/includes/reports_classes.inc

2015-10-18 11:08:15 +0200 Joe Hunt 
Rerun. Fixed wrong calculaiton of standard cost during Advanced Manufacturing. Petros
M	manufacturing/includes/db/work_order_produce_items_db.inc
M	manufacturing/includes/db/work_orders_db.inc

2015-10-17 14:42:42 +0200 Joe Hunt 
Fixed wrong calculaiton of standard cost during Advanced Manufacturing. Petros.
M	manufacturing/includes/db/work_order_produce_items_db.inc
M	manufacturing/includes/db/work_orders_db.inc
M	manufacturing/includes/db/work_orders_quick_db.inc
M	manufacturing/work_order_costs.php

2015-10-15 10:28:09 +0200 Joe Hunt 
Fixed correct calculation of stock_moves standard_cost when producing an Advanced Assembly. Petros
M	manufacturing/includes/db/work_order_produce_items_db.inc
M	manufacturing/includes/db/work_orders_quick_db.inc

2015-10-14 15:34:41 +0200 Joe Hunt 
Changed the GL accounts in Manufacturing to go from Cost accounts to Assembly Accounts. Petros.
M	manufacturing/includes/db/work_order_issues_db.inc
M	manufacturing/includes/db/work_order_produce_items_db.inc
M	manufacturing/includes/db/work_orders_db.inc
M	manufacturing/includes/db/work_orders_quick_db.inc

2015-10-13 20:41:39 +0200 Janusz Dobrowolski 
Fixed net value calculations for GRN leading to unbalanced clearing account.
M	purchasing/includes/db/grn_db.inc
M	purchasing/includes/po_class.inc

2015-10-09 22:47:18 +0200 Joe Hunt 
Material cost doubles itself, if quantity = 0. Fixed.
M	manufacturing/includes/db/work_orders_db.inc

2015-10-06 09:28:40 +0200 Janusz Dobrowolski 
Fixed  deactive_extension hook execution on extension deactivation.
M	admin/inst_module.php
M	includes/hooks.inc

2015-09-07 15:07:27 +0200 Joe Hunt 
Improved safety handling of bad PHP interpretor releases.
M	reporting/includes/html_entity_decode_php4.php

2015-07-11 07:38:04 +0200 Joe Hunt 
Merge branch 'master' of ssh://git.code.sf.net/p/frontaccounting/git
2015-07-11 07:37:36 +0200 Joe Hunt 
New CHANGELOG
M	CHANGELOG.txt

2015-07-10 15:22:28 +0200 Janusz Dobrowolski 
[0003240] Tax Report: duplicated lines were listed for edited transactions.
M	reporting/rep709.php

2015-05-09 23:07:36 +0200 Joe Hunt 
Merge branch 'master' of ssh://git.code.sf.net/p/frontaccounting/git
2015-05-09 23:06:09 +0200 Joe Hunt 
New CHANGELOG.txt format.
M	CHANGELOG.txt
M	doc/CHANGELOG.old.txt

2015-05-06 22:30:47 +0200 Janusz Dobrowolski 

FAMods/js/inserts.js
/**********************************************************************
    Copyright (C) FrontAccounting, LLC.
	Released under the terms of the GNU General Public License, GPL,
	as published by the Free Software Foundation, either version 3
	of the License, or (at your option) any later version.
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
    See the License here <http://www.gnu.org/licenses/gpl-3.0.html>.
***********************************************************************/
var _focus;
var _hotkeys = {
	'alt': false,	// whether is the Alt key pressed
	'list': false, // list of all elements with hotkey used recently
	'focus': -1		// currently selected list element
};

function validate(e) {
	if (e.name && (typeof _validate[e.name] == 'function'))
		return _validate[e.name](e);
	else {
		var n = e.name.indexOf('[');
		if(n!=-1) {
			var key = e.name.substring(n+1, e.name.length-1);
			if (key.length>1 && _validate[e.name.substring(0,n)])
				return _validate[e.name.substring(0,n)][key](e);
		}
	}
	return true;
}

function save_focus(e) {
  _focus = e.name||e.id;
  var h = document.getElementById('hints');
  if (h) {
	h.style.display = e.title && e.title.length ? 'inline' : 'none';
	h.innerHTML = e.title ? e.title : '';
  }
}

function _expand(tabobj) {

  var ul = tabobj.parentNode.parentNode;
  var alltabs=ul.getElementsByTagName("button");
  var frm = tabobj.form;

  if (ul.getAttribute("rel")){
	for (var i=0; i<alltabs.length; i++){
	  alltabs[i].className = "ajaxbutton"  //deselect all tabs // Review CP 2014-11 This will remove all other classes from the element.
	}
	tabobj.className = "current";
	JsHttpRequest.request(tabobj)
  }
}

//interface for selecting a tab (plus expand corresponding content)
function expandtab(tabcontentid, tabnumber) {
  var tabs = document.getElementById(tabcontentid);
 _expand(tabs.getElementsByTagName("input")[tabnumber]);
}

function _set_combo_input(e) {
		e.setAttribute('_last', e.value);
		e.onblur=function() {
		  var but_name = this.name.substring(0, this.name.length-4)+'button';
		  var button = document.getElementsByName(but_name)[0];
		  var select = document.getElementsByName(this.getAttribute('rel'))[0];
		  save_focus(select);
// submit request if there is submit_on_change option set and
// search field has changed.

		  if (button && (this.value != this.getAttribute('_last'))) {
			JsHttpRequest.request(button);
		  } else if(string_contains(this.className, 'combo2')) {
				this.style.display = 'none';
				select.style.display = 'inline';
				setFocus(select);
		  }
		  return false;
		};
		e.onkeyup = function(ev) {
			var select = document.getElementsByName(this.getAttribute('rel'))[0];
			if(select && select.selectedIndex>=0) {
			  var len = select.length;
			  var byid = string_contains(this.className, 'combo') || string_contains(this.className, 'combo3');
			  var ac = this.value.toUpperCase();
			  select.options[select.selectedIndex].selected = false;
			  for (i = 0; i < len; i++) {
				var txt = byid ? select.options[i].value : select.options[i].text;
				if (string_contains(this.className, 'combo3')) {
				  if(txt.toUpperCase().indexOf(ac) == 0) {
					select.options[i].selected = true;
				  	break;
				  }
				} else {
				  if(txt.toUpperCase().indexOf(ac) >= 0) {
					select.options[i].selected = true;
				  	break;
				  }
				}
			  }
			}
		};
    	e.onkeydown = function(ev) {
	  		ev = ev||window.event;
	  		key = ev.keyCode||ev.which;
	  		if(key == 13) {
			  this.blur();
	  		  return false;
	  		}
		}
}

function _update_box(s) {
	var byid = string_contains(s.className, 'combo') || string_contains(s.className, 'combo3');
	var rel = s.getAttribute('rel');
	var box = document.getElementsByName(rel)[0];
		if(box && s.selectedIndex>=0) {
			  var opt = s.options[s.selectedIndex];
				if(box) {
				  var old = box.value;
				  box.value = byid ? opt.value : opt.text;
				  box.setAttribute('_last', box.value);
				  return old != box.value
				}
		}
}

function _set_combo_select(e) {
		// When combo position is changed via js (eg from searchbox)
		// no onchange event is generated. To ensure proper change
		// signaling we must track selectedIndex in onblur handler.
		e.setAttribute('_last', e.selectedIndex);
		e.onblur = function() {
		    var box = document.getElementsByName(this.getAttribute('rel'))[0];
//			if(string_contains(this.className, 'combo'))
//			    _update_box(this);
			if ((this.selectedIndex != this.getAttribute('_last'))
				||((string_contains(this.className, 'combo') || string_contains(this.className, 'combo3')) && _update_box(this))
				)
					this.onchange();
		}
		e.onchange = function() {
			var s = this;
			this.setAttribute('_last', this.selectedIndex);
			if(string_contains(s.className, 'combo') || string_contains(this.className, 'combo3'))
			    _update_box(s);
			if(s.selectedIndex>=0) {
				 var sname = '_'+s.name+'_update';
				 var update = document.getElementsByName(sname)[0];
				 if(update) {
					    JsHttpRequest.request(update);
				}
			}
			return true;
		}
		e.onkeydown = function(event) {
		    event = event||window.event;
		    key = event.keyCode||event.which;
		    var box = document.getElementsByName(this.getAttribute('rel'))[0];
	  		if(key == 8 || (key==37 && event.altKey)) {
				event.returnValue = false;
  			  	return false;
  			}
		    if (box && (key == 32) && (string_contains(this.className, 'combo2'))) {
			    this.style.display = 'none';
			    box.style.display = 'inline';
				box.value='';
				setFocus(box);
			    return false;
			 } else {
			 	if (key == 13 && !e.length) // prevent chrome issue (blocked cursor after CR on empty selector)
			 		return false;
			 }
		}
}

var _w;

function callEditor(key) {
  var el = document.getElementsByName(editors[key][1])[0];
  if(_w) _w.close(); // this is really necessary to have window on top in FF2 :/
  var left = (screen.width - editors[key][2]) / 2;
  var top = (screen.height - editors[key][3]) / 2;
  _w = open(editors[key][0]+el.value+'&popup=1',
	  "edit","scrollbars=yes,resizable=0,width="+editors[key][2]+",height="+editors[key][3]+",left="+left+",top="+top+",screenX="+left+",screenY="+top);
  if (_w.opener == null)
	  _w.opener = self;
  editors._call = key; // store call point for passBack
  _w.focus();
}

function passBack(value) {
	var o = opener;
	if(value != false) {
		var back = o.editors[o.editors._call]; // form input bindings
		var to = o.document.getElementsByName(back[1])[0];
		if (to) {
			if (to[0] != undefined)
				to[0].value = value; // ugly hack to set selector to any value
			to.value = value;
			// update page after item selection
			o.JsHttpRequest.request('_'+to.name+'_update', to.form);
			o.setFocus(to.name);
		}
	}
	close();
}

/*
	Normalize date format using previous input value to guess missing date elements.
	Helps fast date input field change with only single or double numbers (for day or day-and-month fragments)
*/
function fix_date(date, last)
{
	var dat = last.split(user.datesep);
	var cur = date.split(user.datesep);
	var day, month, year;

// TODO: user.date as default?
// TODO: user.datesys
	if (date == "" || date == last) // should we return an empty date or should we return last?
		return date;
	if (user.datefmt == 0 || user.datefmt == 3) // set defaults
	{
		day = dat[1]; month = dat[0]; year = dat[2];
	} else if (user.datefmt == 1 || user.datefmt == 4){
		day = dat[0]; month = dat[1]; year = dat[2];
	} else {
		day = dat[2]; month = dat[1]; year = dat[0];
	}
	if (cur[1] != undefined && cur[1] != "") // day or month entered, could be string 3
	{
		if (user.datefmt == 0 || user.datefmt == 3 || ((user.datefmt == 2 || user.datefmt == 5) && (cur[2] == undefined || cur[2] == "")))
			day = cur[1];
		else
			month = cur[1];
	}
	if (cur[0] != undefined && cur[0] != "") // day or month entered. could be string 3
	{
		if (cur[1] == undefined || cur[1] == "")
			day = cur[0];
		else if (user.datefmt == 0 || user.datefmt == 3 || ((user.datefmt == 2 || user.datefmt == 5) && (cur[2] == undefined || cur[2] == "")))
			month = cur[0];
		else if (user.datefmt == 2 || user.datefmt == 5)
			year = cur[0];
		else
			day = cur[0];
	}
	if (cur[2] != undefined && cur[2] != "") // year,
	{
		if (user.datefmt == 2 || user.datefmt == 5)
			day = cur[2];
		else
			year = cur[2];
	}
	if (user.datefmt<3) {
		if (day<10) day = '0'+parseInt(day, 10);
		if (month<10) month = '0'+parseInt(month, 10);
	}
	if (year<100) year = year<60 ? (2000+parseInt(year,10)) : (1900+parseInt(year,10));

//	console.info(day,month,year)
	if (user.datefmt == 0 || user.datefmt==3)
		return month+user.datesep+day+user.datesep+year;
	if (user.datefmt == 1 || user.datefmt==4)
		return day+user.datesep+month+user.datesep+year;
	return year+user.datesep+month+user.datesep+day;
}

/*
 Behaviour definitions
*/
var inserts = {
	'input': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
				save_focus(this);
				if (string_contains(this.className, 'combo') || string_contains(this.className, 'combo3'))
					this.select();
			};
		}
		if (string_contains(e.className, 'combo') || string_contains(e.className, 'combo2') || string_contains(e.className, 'combo3')) {
				_set_combo_input(e);
		}
		else
    		if(e.type == 'text' ) {
   	  			e.onkeydown = function(ev) {
  					ev = ev||window.event;
  					key = ev.keyCode||ev.which;
 	  				if(key == 13) {
						if(string_contains(e.className, 'searchbox')) e.onblur();
						return false;
					}
					return true;
	  			}
			}
	},
	'input.combo2,input[aspect="fallback"]':
	function(e) {
  	    // this hides search button for js enabled browsers
	    e.style.display = 'none';
	},
	'div.js_only':
	function(e) {
  	    // this shows divs for js enabled browsers only
	    e.style.display = 'block';
	},
	'button': function(e) {
		e.onclick = function(){
			if (validate(e)) {
				setTimeout(function() {	var asp = e.getAttribute('aspect');
					set_mark((asp && (asp.indexOf('process') !== -1)) ? 'progressbar.gif' : 'ajax-loader.gif');
				}, 100);
				return true;
			}
		},
		e.onkeydown = function(ev) {	// block unintentional page escape with 'history back' key pressed on buttons
			ev = ev||window.event;
 			key = ev.keyCode||ev.which;
	  		if(key == 8 || (key==37 && ev.altKey)) {
				ev.returnValue = false;
  			  	return false;
  			}
		}

	},
//	'.ajaxsubmit,.editbutton,.navibutton': // much slower on IE7
	'button.ajaxsubmit,input.ajaxsubmit,input.editbutton,button.editbutton,button.navibutton':
	function(e) {
			e.onclick = function() {
				if (validate(e)) {
					save_focus(e);
					var asp = e.getAttribute('aspect')
					if (asp && (asp.indexOf('process') !== -1))
						JsHttpRequest.request(this, null, 600000); // ten minutes for backup
					else
						JsHttpRequest.request(this);
				}
				return false;
			}
	},
    '.amount': function(e) {
		if(e.onblur==undefined) {
  		  e.onblur = function() {
			var dec = this.getAttribute("dec");
			price_format(this.name, get_amount(this.name), dec);
		  };
		}
	},
	'.searchbox': // emulated onchange event handling for text inputs
		function(e) {
			e.setAttribute('_last_val', e.value);
			e.setAttribute('autocomplete', 'off'); //must be off when calling onblur
  		  	e.onblur = function() {
				var val = this.getAttribute('_last_val');
				if (val != this.value) {
					this.setAttribute('_last_val', this.value);
					JsHttpRequest.request('_'+this.name+'_changed', this.form);
				}
			}
	},
	'.date':
		function(e) {
			e.setAttribute('_last_val', e.value);
			e.setAttribute('autocomplete', 'off');
  		  	e.onblur = function() {
				var val = this.getAttribute('_last_val');
				if (val != this.value) {
					this.value = fix_date(this.value, val);
					this.setAttribute('_last_val', this.value);
					if (e.className.match(/\bactive\b/))
						JsHttpRequest.request('_'+this.name+'_changed', this.form);
				}
			}
	},
	'button[aspect*selector], button[aspect*abort], input[aspect*selector]': function(e) {
		e.onclick = function() {
			passBack(this.getAttribute('rel'));
			return false;
		}
	},
	'button[aspect=popup]': function(e) {
		e.onclick = function() {
			if(_w) _w.close(); // this is really necessary to have window on top in FF2 :/
			 var left = (screen.width - 800)/2;
			 var top = (screen.height - 600)/2;
			  _w = open(document.location+'popup=1',
				  "edit","Scrollbars=0,resizable=0,width=800,height=600, top="+top+",left="+left+",screenX="+left+",screenY="+top);
			  if (_w.opener == null)
				  _w.opener = self;
			//  editors._call = key; // store call point for passBack
//			  _w.moveTo(50, 50);
			  _w.focus();
			return false;
		}
	},
	'select': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
			    save_focus(this);
			};
		}
  		var c = e.className;
		if (string_contains(c, 'combo') || string_contains(c, 'combo2') || string_contains(c, 'combo3'))
			_set_combo_select(e);
		else {
			e.onkeydown = function(ev) {	// block unintentional page escape with 'history back' key pressed on buttons
				ev = ev||window.event;
 				key = ev.keyCode||ev.which;
	  			if(key == 8 || (key=37 && ev.altKey)) {
					ev.returnValue = false;
	  			  	return false;
  				}
			}
		}
	},
	'a.printlink': 	function(l) {
		l.onclick = function() {
		    save_focus(this);
			JsHttpRequest.request(this, null, 60000);
			return false;
		}
	},
	'a.repopts_link': 	function(l) {
		l.onclick = function() {
		    save_focus(this);
		    var replinks = document.getElementsBySelector('a.repopts_link');
				for(var i in replinks)
					replinks[i].style.fontWeight = replinks[i]==this ? 'bold' : 'normal';
			JsHttpRequest.request(this, null);
			return false;
		}
	},
	'a': function(e) { // traverse menu
  		e.onkeydown = function(ev) {
			ev = ev||window.event;
			key = ev.keyCode||ev.which;
			if(key==37 || key==38 || key==39 || key==40) {
					move_focus(key, e, document.links);
					ev.returnValue = false;
					return false;
			}
		}
		// prevent unneeded transaction entry abortion
		if (string_contains(e.className, 'shortcut')
		 || string_contains(e.className, 'menu_option')
		 || string_contains(e.className, 'menu_tab')
 		 || string_contains(e.className, 'selected'))
			e.onclick = function(ev) {
				if (_validate._processing
				 && _validate._modified
				 && !confirm(_validate._processing)) {
					ev.returnValue = false;
					return false;
				}
				if (_hotkeys.alt)	// ommit Chrome accesskeys
					return false;
 				window.location = e.href;
			}
	},
	'ul.ajaxtabs':	function(ul) {
	    var ulist=ul.getElementsByTagName("li");
	    for (var x=0; x<ulist.length; x++){ //loop through each LI e
		var tab=ulist[x].getElementsByTagName("button")[0];
//		if(tab.onclick==undefined) {
// ?  var modifiedurl=ulistlink.getAttribute("href").replace(/^http:\/\/[^\/]+\//i, "http://"+window.location.hostname+"/")
		    var url = tab.form.action
		    tab.onclick=function(){
		    if (!_hotkeys.alt && !this.disabled)
				_expand(this);
			return false;
		    }
		}
//	    }
	},
	'textarea': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
			    save_focus(this);
			};
		}
	}
/*	'tr.editrow': function(e) {
		  	e.onkeydown = function(ev) {
	  		ev = ev||window.event;
	  		key = ev.keyCode||ev.which;
	  		if(key == 13) {
			  // Find & click additem/update button

	  		} else	if(key == 27) {
	  		  return false;
			}
		}

	},
*//*	'#msgbox': function(e) {
	// this is to avoid changing div height after ajax update in IE7
	  e.style.display = e.innerHTML.length ? 'block' : 'none';
	}
*//* TODO
	'a.date_picker':  function(e) {
	    // this un-hides data picker for js enabled browsers
	    e.href = date_picker(this.getAttribute('rel'));
	    e.style.display = '';
	    e.tabindex = -1; // skip in tabbing order
	}
*/
};

function stopEv(ev) {
			if(ev.preventDefault) {
				ev.preventDefault();
				ev.stopPropagation();
			} else {
				ev.returnValue = false;
				ev.cancelBubble = true;
				window.keycode = 0;
			}
			return false;
}
/*
	Modified accesskey system. While Alt key is pressed letter keys moves
	focus to next marked link. Alt key release activates focused link.
*/
function setHotKeys() {
	document.onkeydown = function(ev) {
		ev = ev||window.event;
		key = ev.keyCode||ev.which;
		if (key == 18 && !ev.ctrlKey) {	// start selection, skip Win AltGr
			_hotkeys.alt = true;
			_hotkeys.focus = -1;
			return stopEv(ev);
		}
		else if (ev.altKey && !ev.ctrlKey && ((key>47 && key<58) || (key>64 && key<91))) {
			key = String.fromCharCode(key);
			var n = _hotkeys.focus;
			var l = document.getElementsBySelector('[accesskey='+key+']');
			var cnt = l.length;
			_hotkeys.list = l;
			for (var i=0; i<cnt; i++) {
				n = (n+1)%cnt;
				// check also if the link is visible
				if (l[n].accessKey==key && (l[n].offsetWidth || l[n].offsetHeight)) {
					_hotkeys.focus = n;
	    // The timeout is needed to prevent unpredictable behaviour on IE.
					var tmp = function() {l[_hotkeys.focus].focus();};
					setTimeout(tmp, 0);
					break;
				}
			}
			return stopEv(ev);
		}
		if((ev.ctrlKey && key == 13) || key == 27) {
			_hotkeys.alt = false; // cancel link selection
			_hotkeys.focus = -1;
			ev.cancelBubble = true;
   			if(ev.stopPropagation) ev.stopPropagation();
			// activate submit/escape form
			for(var j=0; j<this.forms.length; j++) {
				var form = this.forms[j];
				for (var i=0; i<form.elements.length; i++){
					var el = form.elements[i];
					var asp = el.getAttribute('aspect');

					if (!string_contains(el.className, 'editbutton') && (asp && asp.indexOf('selector') !== -1) && (key==13 || key==27)) {
						passBack(key==13 ? el.getAttribute('rel') : false);
						ev.returnValue = false;
						return false;
					}
					if (((asp && asp.indexOf('default') !== -1) && key==13)||((asp && asp.indexOf('cancel') !== -1) && key==27)) {
						if (validate(el)) {
							if (asp.indexOf('nonajax') !== -1)
								el.click();
							else
							if (asp.indexOf('process') !== -1)
								JsHttpRequest.request(el, null, 600000);
							else
								JsHttpRequest.request(el);
						}
						ev.returnValue = false;
						return false;
					}
				}
			}
			ev.returnValue = false;
			return false;
		}
		if (editors!=='undefined' && editors[key]) {
			callEditor(key);
			return stopEv(ev); // prevent default binding
		}
		return true;
	};
	document.onkeyup = function(ev) {
		ev = ev||window.event;
		key = ev.keyCode||ev.which;

		if (_hotkeys.alt==true) {
			if (key == 18) {
				_hotkeys.alt = false;
				if (_hotkeys.focus >= 0) {
					var link = _hotkeys.list[_hotkeys.focus];
					if(link.onclick)
						link.onclick();
					else
						if (link.target=='_blank') {
							window.open(link.href,'','toolbar=no,scrollbar=no,resizable=yes,menubar=no,width=900,height=500');
							openWindow(link.href,'_blank');
						} else
							window.location = link.href;
				}
			return stopEv(ev);
			}
		}
		return true;
	}
}

Behaviour.register(inserts);

Behaviour.addLoadEvent(setFocus);
Behaviour.addLoadEvent(setHotKeys);
